content/en/blog/_posts/2023-01-12-protect-mission-critical-pods-priorityclass/index.md

```diff
@@ -0,0 +1,289 @@
+metadata:
+  name: dev-nginx
+  labels:
+    env: dev
+spec:
+  containers:
+  - name: dev-nginx
+    image: nginx
+    resources:
+      requests:
+        memory: "256Mi"
+         cpu: "0.2"
+      limits:
+        memory: ".5Gi"
+        cpu: "0.5"
+```
+
+```yaml
+---
+# preproduction
+apiVersion: v1
+kind: Pod
+metadata:
+  name: preprod-nginx
+  labels:
+    env: preprod
+spec:
+  containers:
+  - name: preprod-nginx
+    image: nginx
+    resources:
+      requests:
+        memory: "1.5Gi"
+        cpu: "1.5"
+      limits:
+        memory: "2Gi"
+        cpu: "2"
+```
+
+```yaml
+---
+# production
+apiVersion: v1
+kind: Pod
+metadata:
+  name: prod-nginx
+  labels:
+    env: prod
+spec:
+  containers:
+  - name: prod-nginx
+    image: nginx
+    resources:
+      requests:
+        memory: "2Gi"
+        cpu: "2"
+      limits:
+        memory: "2Gi"
+        cpu: "2"
+```
+
+Let's create these pods with the `kubectl create -f <FILE.yaml>` command and check their status by using the `kubectl get pods` command to see if they are up and running to serve the traffic.
+
+```console
+$ kubectl get pods --show-labels
+NAME            READY   STATUS    RESTARTS   AGE   LABELS
+dev-nginx       1/1     Running   0          55s   env=dev
+preprod-nginx   1/1     Running   0          55s   env=preprod
+prod-nginx      0/1     Pending   0          55s   env=prod
+```
+
+Bad news. The pod for the Production environment is still Pending, which is bad since it won’t be serving any traffic. The production environment is the one where this matters most.
+
+Let's see why this is happening:
+```console
+$ kubectl get events
+...
+...
+5s          Warning   FailedScheduling   pod/prod-nginx      0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
+```
+
+In this example, there is only one worker node and that node is having some resource crunch.
+
+Now, let's look at how PriorityClass can help in this situation since prod should be given higher priority than the other environments.
+
+## PriorityClass API
+
+Before creating PriorityClasses based on these requirements, let's see what a basic manifest for a PriorityClass looks like and outline some prerequisites:
+
+```yaml
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
+metadata:
+  name: PRIORITYCLASS_NAME
+value: 0 # any integer value between -1000000000 to 1000000000 
+description: >-
+  (Optional) description goes here!
+globalDefault: false # or true. Only one PriorityClass can be the global default.
+```
+
+Below are some prerequisites for PriorityClasses:
+
+- The name of a PriorityClass must be a valid DNS subdomain name.
+- The name should not start with `system-`, as those names are reserved by Kubernetes (for example, they are used for two built-in  PriorityClasses).
+- Its absolute value should be between -1000000000 to 1000000000 (1 billion).
+- Larger numbers are reserved by PriorityClasses such as system-cluster-critical, system-node-critical for kube-system Pods.
```

Comment:
```suggestion
- Larger numbers are reserved by PriorityClasses such as `system-cluster-critical`
  (this Pod is critically important to the cluster) and `system-node-critical` (the node
  critically relies on this Pod).
  `system-node-critical` is a higher priority than `system-cluster-critical`, because a
  cluster-critical Pod can only work well if the node where it is running has all its node-level
  critical requirements met.
```